Sprint #16 Homework

nim0y opened the pull request:
Илья , 
Спасибо тебе большое , что не отправляешь меня всё переписывать ! 
И за то что попытался разгрести всё что я тут наворотил.
Я зарубил себе на носу больше не "бежать впереди паровоза" .
Постарался учесть все твои замечания, но не уверен что сделал всё корректно. 
Еще , раз спасибо за подробное ревью!

nim0y committed:
Sprint #16 -
Most of the changes I have made and commit in sprint #15. Obviously I've made several billions mistakes. I hope it will not become a problem.

nim0y committed:
Sprint #16 fix#1 for visual content
На мой глаз всё совпадает с макетом , но я могу ошибаться.

nim0y committed:
Sprint #16 fix#2 for Player
Не уверен, что ты имел ввиду ,
Но немного переработал VM.
Скорее всего не то сделал , пни в нужную сторону , пожалуйста.

nim0y committed:
Sprint #16 fix#3 for Player
Возможно так правильнее ~~~~~

IliaAnK commented:
Первая итерация правок для AudioPlayerViewModel была ближе к правде.
Ты правильно добавил AudioPlayerState. Именно он и является стейтом экрана, который надо обновлять во ViewModel. 
Но дальше ты свернул немного не туда. Не нужно объединять LiveData-ы с помощью MediatorLiveData. Одной LiveData-ы достаточно.
Я рекомендую откатить код во AudioPlayerViewModel до первой итерации проверки и попробовать поправить на основе следующего примера.

```
// Так будет выглядеть стейт экрана.
data class AudioPlayerState(
    // Здесь не должно быть nullable типов
    val playerState: PlayerState,
    val timerValue: Int
)
```

```
class AudioPlayerViewModel(...) {

    // Единственное поле содержащее стейт
    private val _audioPlayerState = MutableLiveData(
        // Это дефолтный стейт экрана, который будет применён сразу после открытия
        AudioPlayerState(
            playerState = PlayerState.DEFAULT_STATE,
            timerValue = 0
        )
    )

    // Единственное место для подписки
    val audioPlayerState: LiveData<AudioPlayerState> = _audioPlayerState
    
    ...
    
    private val timerRunnable = object : Runnable {
        override fun run() {
            // Обновление стейта
            // С помощью метода "copy" перезаписываем нужное поле data class-а
            // В остальных местах аналогично
            _audioPlayerState.postValue(
                _audioPlayerState.value?.copy(
                    timerValue = playerInteractor.getCurrentPosition()
                )
            )
            handler.postDelayed(this, TIMER_DELAY)
        }
    }
    
    ...
}
```

```
class AudioPlayerActivity : AppCompatActivity() {

    ...

    override fun onCreate(savedInstanceState: Bundle?) {

        ...
        
        // Одно место подписки на единственный стейт
        vm.audioPlayerState.observe(this) { state ->
            binding?.playerTime?.text = SimpleDateFormat("mm:ss", Locale.getDefault()).format(state.timerValue)
            Log.e("myLog", "porisuem")
            execute(state.playerState)
        }
        
        ...
    }
...
}
```

nim0y committed:
Sprint #16 fix#4 for Player
Надеюсь реализовал твои пожелания )

nim0y committed:
Sprint #16 #5 for theme switch ~
Потенциальное решение ,
Обратиться к конфигурации системы.
Не fix  но будем думать .
